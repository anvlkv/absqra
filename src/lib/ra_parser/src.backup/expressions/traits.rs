use super::errors::ParserError;
use ra_lexer::token::{Token, TokenKind};
use ra_lexer::cursor::Position;

pub (crate) trait Expandable<'a, Item, ByItem> {
    fn append_item(self, item: ByItem) -> Result<Item, ParserError>;
    // fn expected_items(&self) -> Vec<TokenKind<'a>>;
}

pub (crate) trait ByTokenExpandableFromRoot<'a, Item> {
    fn append_item(self, token: Token<'a>, depth: Option<u16>) -> Result<Item, ParserError>;
}

pub (crate) trait Leveled {
    fn get_level(&self) -> u16;
}

pub (crate) trait Positioned {
    fn get_position(&self) -> (Position, Position);
}