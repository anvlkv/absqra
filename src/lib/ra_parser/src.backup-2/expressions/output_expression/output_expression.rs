use super::*;
use serde::ser::{Serialize,Serializer};
use std::convert::TryInto;


#[derive(Debug, Clone, Serialize)]
pub struct OutputExpression<'a> (
    pub OutputExpressionMember<'a>, 
    pub Option<Box<OutputExpression<'a>>>
);

impl<'a> Serialize for OutputExpressionMember<'a> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self.clone() {
            Self::Buffer(v) => {
                let count: u64 = v.len().try_into().unwrap();
                serializer.serialize_u64(count)
            }
            Self::Identifier(token) => token.serialize(serializer),
            Self::Literal(token) => token.serialize(serializer),
            // Self::Group(exp) => exp.serialize(serializer),
            // Self::Operation(exp) => exp.serialize(serializer),
        }
    }
}

impl <'a> ParsedByToken<'a> for OutputExpression<'a> {
    fn new_from_token(token: RaToken<'a>) -> Result<Box<Self>, Vec<ParserError>> { todo!() }
    fn append_token(self, token: RaToken<'a>) -> Result<Box<Self>, Vec<ParserError>> { todo!() }
    fn allowed_tokens(&self) -> Vec<TokenKind<'a>> { todo!() }
    fn required_tokens(&self) -> Vec<TokenKind<'a>> { todo!() }
}

impl <'a> StartsWithTokens<'a> for OutputExpression<'a> {
    fn starts_with_tokens() -> Vec<TokenKind<'a>> { 
        OutputExpressionMember::starts_with_tokens()
    }
}