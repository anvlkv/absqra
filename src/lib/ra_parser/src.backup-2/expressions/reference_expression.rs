use super::*;

#[derive(Serialize, Clone, Debug)]
pub struct ReferenceExpression {

}

impl<'a> ParsedByToken<'a> for ReferenceExpression {
    fn new_from_token(token: RaToken<'a>) -> Result<Box<ReferenceExpression>, Vec<ParserError>> { 
        todo!("implement new")
    }

    fn append_token(self, token: RaToken<'a>) -> Result<Box<ReferenceExpression>, Vec<ParserError>> { 
        todo!("implement append_token")
    }
    
    fn allowed_tokens(&self) -> Vec<TokenKind<'a>> { 
        todo!("implement allowed_tokens")
    }

    fn required_tokens(&self) -> Vec<TokenKind<'a>> { 
        todo!() 
    }
}

impl<'a> StartsWithTokens<'a> for ReferenceExpression {
    fn starts_with_tokens() -> Vec<TokenKind<'static>> { 
        vec![
            TokenKind::At,
            TokenKind::Identifier(Default::default())
        ]
     }
}

