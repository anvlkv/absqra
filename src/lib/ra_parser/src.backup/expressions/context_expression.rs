use super::errors::ParserError;
use super::traits::*;
use failure::Backtrace;
use ra_lexer::cursor::Position;
use ra_lexer::token::{Token, TokenKind};
use serde::Serialize;

#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum ContextExpressionMemberKind {
    None,
    One,
    N,
    MSpecifier,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum ContextExpressionMember {
    Target(ContextExpressionMemberKind),
    Source(ContextExpressionMemberKind),
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct ContextExpression(pub ContextExpressionMember, pub Option<ContextExpressionMember>);

impl<'a> ContextExpression {
    pub fn new(token: Token<'a>) -> Result<Self, ParserError> {
        // println!("{:?}", token);
        match token.kind.unwrap() {
            TokenKind::Int(1) => Ok(ContextExpression(
                ContextExpressionMember::Target(ContextExpressionMemberKind::One),
                None,
            )),
            TokenKind::Identifier("N") => Ok(ContextExpression(
                ContextExpressionMember::Target(ContextExpressionMemberKind::N),
                None,
            )),
            TokenKind::OpenCurlyBrace => Ok(ContextExpression(
                ContextExpressionMember::Target(ContextExpressionMemberKind::MSpecifier),
                None,
            )),
            _ => Err(ParserError::ExpectedAGotB(
                format!("{}", token),
                format!(
                    "{:?}",
                    vec![
                        TokenKind::Int(1),
                        TokenKind::Identifier("N"),
                        TokenKind::OpenCurlyBrace
                    ]
                ),
                token.position.0,
                Backtrace::new(),
            )),
        }
    }
}

impl Leveled for ContextExpression {
    fn get_level(&self) -> u16 {
        todo!()
    }
}

impl Positioned for ContextExpression {
    fn get_position(&self) -> (Position, Position) {
        todo!()
    }
}

impl<'a> Expandable<'a, ContextExpression, Token<'a>> for ContextExpression {
    fn append_item(self, token: Token<'a>) -> Result<ContextExpression, ParserError> {
        match token.kind.unwrap() {
            TokenKind::Colon => Ok(ContextExpression(
                self.0.clone(),
                Some(ContextExpressionMember::Source(
                    ContextExpressionMemberKind::None,
                )),
            )),
            _ => {
                if self.1.is_some() {
                    match token.kind.unwrap() {
                        TokenKind::Int(1) => Ok(ContextExpression(
                            self.0.clone(),
                            Some(ContextExpressionMember::Source(
                                ContextExpressionMemberKind::One,
                            )),
                        )),
                        TokenKind::Identifier("N") => Ok(ContextExpression(
                            self.0.clone(),
                            Some(ContextExpressionMember::Source(
                                ContextExpressionMemberKind::N,
                            )),
                        )),
                        TokenKind::OpenCurlyBrace => Ok(ContextExpression(
                            self.0.clone(),
                            Some(ContextExpressionMember::Source(
                                ContextExpressionMemberKind::MSpecifier,
                            )),
                        )),
                        TokenKind::CloseCurlyBrace => Ok(self.clone()),
                        _ => Err(ParserError::ExpectedAGotB(
                            format!("{}", token),
                            format!(
                                "{:?}",
                                vec![
                                    TokenKind::Int(1),
                                    TokenKind::Identifier("N"),
                                    TokenKind::OpenCurlyBrace,
                                    TokenKind::CloseCurlyBrace
                                ]
                            ),
                            token.position.0,
                            Backtrace::new(),
                        )),
                    }
                } else {
                    Err(ParserError::ExpectedAGotB(
                        format!("{}", token),
                        format!("{:?}", vec![TokenKind::Colon]),
                        token.position.0,
                        Backtrace::new(),
                    ))
                }
            }
        }
    }
}
