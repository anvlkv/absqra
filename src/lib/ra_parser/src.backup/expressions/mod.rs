pub mod context_expression;
pub mod output_expression;
pub mod reference_expression;
pub mod input_expression;
pub mod annotation_expression;
pub mod content;
pub (crate) mod traits;

pub (crate) use super::errors;