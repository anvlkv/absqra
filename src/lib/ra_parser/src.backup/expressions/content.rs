use ra_lexer::cursor::{Cursor, Position};
use ra_lexer::token::{RaToken, TokenKind};
use ra_lexer::tokenize;
use serde::Serialize;

use super::reference_expression::ReferenceExpression;
use super::traits::Expandable;
use failure::Backtrace;
use super::errors::ParserError;

#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum ContentBlockMember<'a> {
    Body(String),
    Template(ReferenceExpression<'a>),
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct Content<'a>(ContentBlockMember<'a>, Option<Box<Content<'a>>>);

impl<'a> Content<'a> {
    pub fn new(token: RaToken<'a>) -> Result<Content<'a>, ParserError> {
        match token.kind.unwrap() {
            TokenKind::ContentBlock => Self::parse_content(token),
            _ => Err(ParserError::ExpectedAGotB(
                format!("{}", token),
                format!("{:?}", vec![TokenKind::ContentBlock]),
                token.position.0,
                Backtrace::new()
            )),
        }
    }

    fn parse_content(token: RaToken<'a>) -> Result<Content<'a>, ParserError> {
        let RaToken {
            kind,
            level,
            len,
            content,
            position,
        } = token;

        let mut cursor = Cursor::new(content, position.0, level, 0);

        let mut content = Content(Self::parse_member(&mut cursor)?, None);

        while !cursor.is_eof() {
            match Self::parse_member(&mut cursor) {
                Ok(member) => {
                    content = content.append_item(member)?;
                }
                Err(e) => return Err(e),
            }
        }

        Ok(content)
    }

    fn parse_member(cursor: &mut Cursor<'a>) -> Result<ContentBlockMember<'a>, ParserError> {
        let mut string_buffer = String::new();
        let mut token_buffer = cursor.slice(0, 0);

        while let Some(ch) = cursor.bump() {
            match ch {
                '{' => {
                    match cursor.first_ahead() {
                        '{' => {
                            if string_buffer.len() > 0 {
                                return Ok(ContentBlockMember::Body(string_buffer));
                            }
                            let start_consumed = cursor.len_consumed();
                            while let Some(ch) = cursor.bump() {
                                match ch {
                                    '}' => match cursor.first_ahead() {
                                        '}' => {
                                            if token_buffer.len() > 0 {
                                                return Ok(ContentBlockMember::Template(
                                                    Self::parse_reference_from_buffer(
                                                        token_buffer,
                                                        cursor.position,
                                                    )?,
                                                ));
                                            }
                                            break;
                                        }
                                        _ => return Err(ParserError::InvalidBlock(Backtrace::new())),
                                    },
                                    _ => {
                                        token_buffer =
                                            cursor.slice(start_consumed, cursor.len_consumed())
                                        // token_buffer.push(ch)
                                    }
                                }
                            }
                        }
                        '\\' => {
                            cursor.bump();
                        }
                        _ => {
                            string_buffer.push(ch);
                        }
                    }
                }
                _ => {
                    string_buffer.push(ch);
                }
            }
        }

        if token_buffer.len() > 0 {
            Err(ParserError::UnexpectedEndOfInput(cursor.position, Backtrace::new()))
        } else {
            Ok(ContentBlockMember::Body(string_buffer))
        }
    }

    fn parse_reference_from_buffer(
        buffer: &'a str,
        position: Position,
    ) -> Result<ReferenceExpression<'a>, ParserError> {
        let mut tokens_stream = tokenize(buffer);

        match tokens_stream.next() {
            Some(token) => {
                let mut expression = ReferenceExpression::new(token?)?;

                while let Some(token) = tokens_stream.next() {
                    expression = expression.append_item(token?)?;
                }

                Ok(expression)
            }
            None => Err(ParserError::UnexpectedEndOfInput(position, Backtrace::new())),
        }
    }
}

impl<'a> Expandable<'a, Content<'a>, ContentBlockMember<'a>> for Content<'a> {
    fn append_item(self, item: ContentBlockMember<'a>) -> Result<Content<'a>, ParserError> {
        let Content(current, next) = self;
        match next {
            Some(next) => Ok(Content(current, Some(Box::new(next.append_item(item)?)))),
            None => Ok(Content(current, Some(Box::new(Content(item, None))))),
        }
    }
}
