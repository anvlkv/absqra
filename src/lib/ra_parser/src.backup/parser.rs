use crate::expressions::context_expression::{ContextExpression, ContextExpressionMemberKind, ContextExpressionMember};
use ra_lexer::token::{RaToken, TokenKind};
use ra_lexer::tokenize;
use ra_lexer::errors::LexerError;

use super::block::{Block, BlockKind};
use super::cursor::Cursor;
use super::errors::ParserError;
use super::expressions::traits::{Leveled, Expandable, Positioned};
use failure::Backtrace;

pub fn parse<'a>(input: &'a str) -> Result<Block<'a>, (Vec<ParserError>, Block<'a>)> {
    let stream = tokenize(input);
    
    let mut errors: Vec<ParserError> = Vec::new();

    let program = {
        let mut block = Block::new(RaToken::default()).unwrap();

        let mut tokens_cursor = Cursor::new(
            stream
            .take_while(|r| r.is_ok())
            .map(|r| r.unwrap())
            .filter(|tok| tok.kind.unwrap() != TokenKind::Comment)
        );
        loop {
            if !tokens_cursor.is_eof() {
                match tokens_cursor.advance_block() {
                    Ok(blk) => block.children.push(blk),
                    Err(es) => errors.extend(es.into_iter()),
                }
            } else {
                break;
            }
        }

        block
    };

    if errors.len() > 0 {
        Err((errors, program))
    } else {
        Ok(program)
    }
}

impl<'token, I> Cursor<'token, I>
where
    I: Iterator<Item = RaToken<'token>>,
{
    fn advance_block<'a>(&'a mut self) -> Result<Block<'token>, Vec<ParserError>> {
        let first_token = self.bump();
        let mut errors: Vec<ParserError> = Vec::new();
        match first_token {
            Some(token) => {
                self.parse_block(token)
            },
            None => {
                errors.push(ParserError::UnexpectedEndOfInput(self.position, Backtrace::new()));
                Err(errors)
            }
        }
    }

    fn parse_block(
        &mut self,
        first_token: RaToken<'token>,
    ) -> Result<Block<'token>, Vec<ParserError>> {
        let mut errors: Vec<ParserError> = Vec::new();
        let block = Block::new(first_token)?;

        if let (Some(blk), mut errs) =
            self.check_parse_block_expression(block, &mut errors)
        {
            if let (Some(blk), mut errs) = self.check_parse_block_children(blk, &mut errs) {
                if let (Some(blk), errs) = self.check_parse_union(blk, &mut errs) {
                    if errs.len() == 0 {
                        return Ok(blk);
                    }
                }
            }
        }

        Err(errors)
    }

    fn check_parse_block_expression<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut lvl = block.get_level();
        let mut line_number = (block.get_position().0).0;
        while self.first_ahead().is_some()
            && self.first_ahead().unwrap().level == lvl
            && (self.first_ahead().unwrap().position.0).0 == line_number {
                match block.clone().append_item(self.bump().unwrap()) {
                    Ok(blk) => {
                        block = blk;

                        // // TODO: ugly, move this logic elsewhere

                        if let BlockKind::ContextModification(Some(ContextExpression(target, source))) = block.clone().kind { 
                            if ContextExpressionMember::Target(ContextExpressionMemberKind::MSpecifier) == target && source.is_none() {
                                if let (Some(mut blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    let close_curly_brace = self.bump().unwrap();
                                    assert_eq!(close_curly_brace.kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    lvl = close_curly_brace.level;
                                    line_number = (close_curly_brace.position.0).0;
                            //         // let updated_expression = ContextExpression(target, None).append_item(self.bump().unwrap())?;

                            //         // blk.kind = BlockKind::ContextModification(Some(updated_expression));
                            //         // self.check_parse_block_expression(blk, errs)
                                }
                            }
                            else if source.is_some() && ContextExpressionMember::Source(ContextExpressionMemberKind::MSpecifier) == source.unwrap() {
                                if let (Some(blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    // return Ok(blk);
                                }
                            }
                            println!("{:?}", block);
                        }

                        // // end TODO
                    },
                    Err(e) => {
                        errors.push(e)
                    }
                }
        }

        (Some(block), errors)
    }

    fn check_parse_block_children<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        
        while !self.is_eof() && self.first_ahead().unwrap().level == block.get_level() + 1 {
            match self.advance_block() {
                Ok(child) => {
                    block.children.push(child);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }

        if errors.len() == 0 {
            (Some(block), errors)
        } else {
            (None, errors)
        }
    }

    fn check_parse_union<'errors>(
        &mut self,
        block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut union_block = block.clone();
        let lvl = block.get_level();
        let mut union_size = 0;
        while !self.is_eof()
            && self.first_ahead().unwrap().kind.unwrap() == TokenKind::Coma
            && self.first_ahead().unwrap().level == lvl
        {
            assert!(self.bump().unwrap().kind.unwrap() == TokenKind::Coma);
            match self.advance_block() {
                Ok(blk) => {
                    union_size += 1;
                    union_block.children.push(blk);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }
        
        if union_size > 0 {
            union_block.kind = BlockKind::Union(union_size);
            (Some(union_block), errors)
        } else {
            (Some(block), errors)
        }
    }
}
