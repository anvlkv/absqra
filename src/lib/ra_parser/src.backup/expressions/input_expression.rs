use super::content::Content;
use super::errors::ParserError;
use super::output_expression::OutputExpression;
use super::reference_expression::ReferenceExpression;
use super::traits::*;
use serde::Serialize;
use failure::Backtrace;

use ra_lexer::cursor::Position;
use ra_lexer::token::{Token, TokenKind};

#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum ArgumentType<'a> {
    Named(Token<'a>, bool),
    Ordered(u16),
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum ValueType<'a> {
    Literal(Token<'a>),
    Content(Content<'a>),
    OutputExpression(OutputExpression<'a>),
    ReferenceExpression(ReferenceExpression<'a>),
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct InputExpression<'a>(
    Option<ArgumentType<'a>>,
    Option<ValueType<'a>>,
    Option<Box<InputExpression<'a>>>,
);

impl<'a> InputExpression<'a> {
    pub fn new(token: Token<'a>) -> Result<Self, ParserError> {
        match token.kind.unwrap() {
            TokenKind::Colon => Ok(Self(None, None, None)),
            _ => Err(ParserError::ExpectedAGotB(
                format!("{}", token),
                format!("{:?}", vec![TokenKind::Colon]),
                token.position.0,
                Backtrace::new()
            )),
        }
    }
}

impl<'a> Leveled for InputExpression<'a> {
    fn get_level(&self) -> u16 {
        let InputExpression(argument_type, value_type, next) = self;
        // isf argument_type.is_some() {
        // argument_type.unwrap().
        // }
        todo!()
    }
}

impl<'a> Positioned for InputExpression<'a> {
    fn get_position(&self) -> (Position, Position) {
        let InputExpression(argument_type, value_type, next) = self;
        todo!()
    }
}

impl<'a> ByTokenExpandableFromRoot<'a, InputExpression<'a>> for InputExpression<'a> {
    fn append_item(
        self,
        token: Token<'a>,
        depth: Option<u16>,
    ) -> Result<InputExpression<'a>, ParserError> {
        let InputExpression(argument_type, value_type, next) = self;

        let ordered_argument_depth = match depth {
            Some(d) => d,
            None => 0,
        };

        if argument_type.is_none() {
            match token.kind.unwrap() {
                TokenKind::Identifier(_) => Ok(InputExpression(
                    Some(ArgumentType::Named(token, false)),
                    None,
                    None,
                )),
                TokenKind::Equals => Ok(InputExpression(
                    Some(ArgumentType::Ordered(ordered_argument_depth)),
                    None,
                    None,
                )),
                _ => Err(ParserError::ExpectedAGotB(
                    format!("{}", token),
                    format!("{:?}", vec![TokenKind::Identifier(""), TokenKind::Equals]),
                    token.position.0,
                    Backtrace::new()
                )),
            }
        } else if value_type.is_none() {
            match argument_type.clone().unwrap() {
                ArgumentType::Named(tok, assigned) => {
                    if !assigned {
                        return match token.kind.unwrap() {
                            TokenKind::Equals => Ok(InputExpression(
                                Some(ArgumentType::Named(tok, true)),
                                None,
                                None,
                            )),
                            _ => {

                                Ok(InputExpression(
                                    Some(ArgumentType::Ordered(ordered_argument_depth)),
                                    Some(ValueType::ReferenceExpression(ReferenceExpression::new(tok)?.append_item(token)?)),
                                    None
                                ))
                            }
                        };
                    }
                }
                _ => {}
            }

            match token.kind.unwrap() {
                TokenKind::StringLiteral(_) | TokenKind::Float(_) | TokenKind::Int(_) => Ok(
                    InputExpression(argument_type, Some(ValueType::Literal(token)), None),
                ),
                TokenKind::ContentBlock => {
                    let content = Content::new(token)?;
                    Ok(InputExpression(
                        argument_type,
                        Some(ValueType::Content(content)),
                        None,
                    ))
                }
                TokenKind::Identifier(_) => Ok(InputExpression(
                    argument_type,
                    Some(ValueType::ReferenceExpression(ReferenceExpression::new(
                        token,
                    )?)),
                    None,
                )),
                _ => {
                    match OutputExpression::new(token) {
                        Ok(expression) => {
                            Ok(InputExpression(
                                argument_type,
                                Some(ValueType::OutputExpression(expression)),
                                None
                            ))
                        },
                        Err(e) => {
                            Err(ParserError::ChainedError(
                                Box::new(e),
                                Box::new(ParserError::ExpectedAGotB(
                                    format!("{}", token),
                                    format!(
                                        "{:?}",
                                        vec![
                                            TokenKind::StringLiteral(""),
                                            TokenKind::Float(0.0),
                                            TokenKind::Int(0),
                                            TokenKind::Identifier("")
                                        ]
                                    ),
                                    token.position.0,
                                    Backtrace::new()
                                ))
                            ))               
                        }
                    }
                },
            }
        } else if next.is_none() {
            match value_type.clone().unwrap() {
                ValueType::ReferenceExpression(expression) => {
                    let next_expression = match expression.clone().append_item(token) {
                        Ok(expression) => expression,
                        Err(e) => {
                            match e {
                                ParserError::ExpectedAGotB(A, B, position, trace) => {
                                    let mut next_expression: OutputExpression = expression.into();
                                    next_expression = next_expression.append_item(token)?;
                                    return Ok(InputExpression(
                                        argument_type,
                                        Some(ValueType::OutputExpression(next_expression)),
                                        None
                                    ))
                                },
                                _ => {
                                    return Err(e);
                                }
                            }
                        }
                    };
                    Ok(InputExpression(
                        argument_type,
                        Some(ValueType::ReferenceExpression(next_expression)),
                        None
                    ))
                },
                ValueType::OutputExpression(expression) => {
                    let next_expression = expression.append_item(token)?;
                    Ok(InputExpression(
                        argument_type,
                        Some(ValueType::OutputExpression(next_expression)),
                        None
                    ))
                }
                _ => {
                    let next_expression = InputExpression::new(token)?;
                    Ok(InputExpression(
                        argument_type,
                        value_type,
                        Some(Box::new(next_expression)),
                    ))
                }
            }
        } else {
            let updated_expression = next
                .unwrap()
                .append_item(token, Some(ordered_argument_depth + 1))?;
            Ok(InputExpression(
                argument_type,
                value_type,
                Some(Box::new(updated_expression)),
            ))
        }
    }
}
