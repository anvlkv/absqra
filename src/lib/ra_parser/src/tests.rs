mod lib {
    use ra_lexer::cursor::Position;
    use ra_lexer::token::{Token, TokenKind};

    use crate::block::BlockKind;
    use crate::expressions::output_expression::ExpressionMember;
    use crate::expressions::output_expression::{MathOperation, OperationKind, OutputExpression};
    use crate::expressions::reference_expression::ReferenceExpression;
    use crate::parser::parse;

    #[test]
    fn it_should_return_program_block() {
        assert_eq!(parse("abc").expect("can't parse").kind, BlockKind::Program);
    }

    #[test]
    fn it_should_parse_block_level_expressions() {
        let program = parse("abc + 2").expect("can't parse");
        let parsed = program.children.iter().nth(0).unwrap();
        assert_eq!(
            parsed.kind,
            BlockKind::Output(OutputExpression(
                Box::new(ExpressionMember::ReferenceExpression(ReferenceExpression(Token {
                    kind: Some(TokenKind::Identifier("abc")),
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    content: "abc",
                    level: 0,
                }, None))),
                Some(OperationKind::MathOperation(MathOperation::Sum)),
                Some(Box::new(ExpressionMember::Literal(Token {
                    kind: Some(TokenKind::Int(2)),
                    position: (Position(1, 6), Position(1, 7)),
                    level: 0,
                    content: "2",
                    len: 1
                })))
            ))
        )
        // assert_eq!(parsed.expression.1, Some(OperationKind::MathOperation(MathOperation::Sum)));

        // assert_eq!(
        //     parsed.expression.0,
        //     Token {
        //         content: "abc",
        //         kind: Some(TokenKind::Identifier("abc")),
        //         len: 3,
        //         level: 0,
        //         position: (Position(1, 0), Position(1, 3))
        //     }
        // );

        // assert_eq!(
        //     (parsed.expression.2.as_ref().unwrap())
        //     Token {
        //         content: "2",
        //         kind: Some(TokenKind::Int(2)),
        //         len: 1,
        //         level: 0,
        //         position: (Position(1, 6), Position(1, 7))
        //     }
        // );
    }

    use ra_dev_tools::insta::{assert_json_snapshot, assert_snapshot};
    
    use std::fs::File;
    use std::io::Read;
    use ra_dev_tools::for_each_ra_example_file;

    #[test]
    fn it_should_match_snapshots() {
        println!("{:?}", std::env::args());
        for_each_ra_example_file(|example| {
            let mut file = File::open(example.path()).unwrap();
            
            let mut contents = String::new();
            file.read_to_string(&mut contents).unwrap();

            let block_tree = {
                match parse(&contents) {
                    Ok(b) => b,
                    Err((errors, parsed)) => {

                        let formatted_errors: String = {
                            errors.into_iter().map(|e| format!("{}", e)).collect::<Vec<String>>().join("\n")
                        };

                        
                        assert_snapshot!(String::from(example.path().to_str().unwrap()) + "__ERR__MESSAGES__", formatted_errors);
                        assert_json_snapshot!(String::from(example.path().to_str().unwrap()) + "__ERR__", parsed);


                        panic!("failed to parse example {:?}", example.path());
                    }
                }
            };
            assert_json_snapshot!(example.path().to_str().unwrap(), block_tree)
        })       
    }
}
