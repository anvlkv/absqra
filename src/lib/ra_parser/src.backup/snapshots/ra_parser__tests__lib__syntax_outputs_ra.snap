---
source: src/lib/ra_parser/src/tests.rs
expression: block_tree
---
{
  "kind": "Program",
  "children": [],
  "first_token": {
    "kind": null,
    "len": 1,
    "content": "",
    "position": [
      [
        0,
        0
      ],
      [
        0,
        1
      ]
    ],
    "level": 0
  }
}
